### ISSUE TITLE
Integrate with express somehow

### ISSUE THREAD

**Issue by wmertens** [montagejs/joey#13]
There is a large ExpressJS ecosystem, and there are many interesting middlewares like Passport for authentication.

I'd like to discuss the possibilities of integrating the two.

I see two possibilities:
- an adaptor that provides expressjs-compatible `req` and `res` objects
- an Express middleware that runs a Joey chain

On the face of it, the latter seems the most plausible since it means not running after Express trying to copy its API.

Does that sound plausible/am I on crack? Any tips for implementing?

**Comment by Stuk** [comment]
Yeah, I've thought about this idea, exactly because there's a large ecosystem of Express packages.

I briefly thought about making Express compatible request/response objects, but (as far as I can remember) there are several big compat issues, such as Joey's body streaming being very different to the built in streams.

An Express middleware that runs Joey chains sounds like an interesting idea that's worth exploring. The main thing to think about would be how the Joey response object could be translated to the Express response object.

**Comment by wmertens** [comment]
@Stuk I'm thinking that basically there would be a `.express()` instead of a `.server()` and it would generate Express middleware that runs the request through the chain and if the response is not `undefined` it `res.send()`s it. I suppose errors should just be sent to next().

There might be Joey apps that aren't happy about getting a pre-processed request object, like the bodyparser, but in general it should just work methinks.

Hmmm I should just try it, see if you can beat me to it :grin:.

**Comment by Stuk** [comment]
I'm not going to have any time to take a look soon, so please, go ahead!

**Comment by kriskowal** [comment]
`.express()` or `.connect()` (from which Express gets its protocol) seems sane. Seems analogous to `promise.nodeify()`. Could reuse `nodeify()` since it might as well be the Node.js convention.
